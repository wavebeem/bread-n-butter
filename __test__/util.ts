import * as bnb from "../src/bread-n-butter";

export function snapTest<A>(parser: bnb.Parser<A>, input: string): void {
  expect(parser.parse(input)).toMatchSnapshot(JSON.stringify(input));
}
