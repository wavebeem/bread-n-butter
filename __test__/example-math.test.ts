import SimpleMath, {
  MathOperator1,
  MathOperator2,
  MathExpr,
} from "../examples/math";
import { snapTest } from "./util";

test("2 part expression", () => {
  const text = `2 * 3 + 4`;
  const expr = SimpleMath.tryParse(text);
  snapTest(SimpleMath, text);
  expect(expr.calculate()).toEqual(10);
  expect(expr.toString()).toEqual(`((2 * 3) + 4)`);
});

test("2 part expression with parentheses", () => {
  const text = `2 * (3 + 4)`;
  const expr = SimpleMath.tryParse(text);
  snapTest(SimpleMath, text);
  expect(expr.calculate()).toEqual(14);
  expect(expr.toString()).toEqual(`(2 * (3 + 4))`);
});

test("unary and binary minus", () => {
  const text = `-1 - -1`;
  const expr = SimpleMath.tryParse(text);
  snapTest(SimpleMath, text);
  expect(expr.calculate()).toEqual(0);
  expect(expr.toString()).toEqual(`((- 1) - (- 1))`);
});

test("large expression", () => {
  const text = `-2 + 3 * 4 - 5 / 7 ** 6`;
  const expr = SimpleMath.tryParse(text);
  snapTest(SimpleMath, text);
  expect(expr.calculate()).toEqual(9.999957500701239);
});

test("MathOperator1", () => {
  const expr: MathExpr = {
    calculate() {
      return 0;
    },
  };
  expect(() => new MathOperator1("bad", expr).calculate()).toThrow();
});

test("MathOperator2", () => {
  const expr: MathExpr = {
    calculate() {
      return 0;
    },
  };
  expect(() => new MathOperator2("bad", expr, expr).calculate()).toThrow();
});
