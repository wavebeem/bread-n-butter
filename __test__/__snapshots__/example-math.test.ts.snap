// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`2 part expression with parentheses: "2 * (3 + 4)" 1`] = `
Object {
  "type": "ParseOK",
  "value": MathOperator2 {
    "left": MathNumber {
      "value": 2,
    },
    "operator": "*",
    "right": MathOperator2 {
      "left": MathNumber {
        "value": 3,
      },
      "operator": "+",
      "right": MathNumber {
        "value": 4,
      },
    },
  },
}
`;

exports[`2 part expression: "2 * 3 + 4" 1`] = `
Object {
  "type": "ParseOK",
  "value": MathOperator2 {
    "left": MathOperator2 {
      "left": MathNumber {
        "value": 2,
      },
      "operator": "*",
      "right": MathNumber {
        "value": 3,
      },
    },
    "operator": "+",
    "right": MathNumber {
      "value": 4,
    },
  },
}
`;

exports[`large expression: "-2 + 3 * 4 - 5 / 7 ** 6" 1`] = `
Object {
  "type": "ParseOK",
  "value": MathOperator2 {
    "left": MathOperator2 {
      "left": MathOperator1 {
        "expression": MathNumber {
          "value": 2,
        },
        "operator": "-",
      },
      "operator": "+",
      "right": MathOperator2 {
        "left": MathNumber {
          "value": 3,
        },
        "operator": "*",
        "right": MathNumber {
          "value": 4,
        },
      },
    },
    "operator": "-",
    "right": MathOperator2 {
      "left": MathNumber {
        "value": 5,
      },
      "operator": "/",
      "right": MathOperator2 {
        "left": MathNumber {
          "value": 7,
        },
        "operator": "**",
        "right": MathNumber {
          "value": 6,
        },
      },
    },
  },
}
`;

exports[`unary and binary minus: "-1 - -1" 1`] = `
Object {
  "type": "ParseOK",
  "value": MathOperator2 {
    "left": MathOperator1 {
      "expression": MathNumber {
        "value": 1,
      },
      "operator": "-",
    },
    "operator": "-",
    "right": MathOperator1 {
      "expression": MathNumber {
        "value": 1,
      },
      "operator": "-",
    },
  },
}
`;
